File: output/StarCoder/codenet/C++/Python/s023282516.py
import sys

S = sys.stdin.readline().strip()

# 後ろから解くかわりにすべての文字列を「左右反転」する
S = S[::-1]
for i in range(4):
    divide[i] = divide[i][::-1]

# 端から切っていく
can = True
for i in range(len(S)):
    can2 = False
    for j in range(4):
        d = divide[j]
        if S[i:i+len(d)] == d:
            can2 = True
            i += len(d)
    if not can2:
        can = False
        break

if can:
    print("YES")
else:
    print("NO")

File: output/StarCoder/codenet/C++/Java/s023282516.java
import java.util.*;

public class s023282516 {

    public static void main(String[] args) {
        Scanner sc = new Scanner(System.in);
        String S = sc.next();

        // 後ろから解くかわりにすべての文字列を「左右反転」する
        reverse(S.begin(), S.end());
        String[] divide = {"dream", "dreamer", "erase", "eraser"};
        for (int i = 0; i < 4; ++i) reverse(divide[i].begin(), divide[i].end());

        // 端から切っていく
        boolean can = true;
        for (int i = 0; i < S.size();) {
            boolean can2 = false; // 4 個の文字列たちどれかで divide できるか
            for (int j = 0; j < 4; ++j) {
                String d = divide[j];
                if (S.substring(i, i + d.length()).equals(d)) { // d で divide できるか
                    can2 = true;
                    i += d.length(); // divide できたら i を進める
                }
            }
            if (!can2) { // divide できなかったら
                can = false;
                break;
            }
        }

        if (can) System.out.println("YES");
        else System.out.println("NO");
    }
}